src/logger/src/logger_impl.cpp

```diff
@@ -0,0 +1,40 @@
+#include "logger/logger_impl.h"
+#include <bsp/ui.h>
+#include <cstdarg>
+
+LoggerImpl::LoggerImpl(LogLevel level, UI* ui) {
+    m_log_level = level;
+    m_ui = ui;
```

Comment:
We should define our standard on null checks if we decide to do them, there is one missing here

Reply:
We could maybe use class references instead of pointers when possible
https://isocpp.org/wiki/faq/references

Reply:
In this case, the UI class should be a singleton. We should manage it using an std:shared_pointer. This way, we don't have to do a null check on the parameter passed since it would be caught on the creation of the shared_pointer. Having a shared pointer would also add protection for multithread use. It has relatively low overhead for embedded systems (https://stackoverflow.com/questions/22295665/how-much-is-the-overhead-of-smart-pointers-compared-to-normal-pointers-in-c).

Reply:
> In this case, the UI class should be a singleton. We should manage it using an std:shared_pointer. This way, we don't have to do a null check on the parameter passed since it would be caught on the creation of the shared_pointer. Having a shared pointer would also add protection for multithread use. It has relatively low overhead for embedded systems (https://stackoverflow.com/questions/22295665/how-much-is-the-overhead-of-smart-pointers-compared-to-normal-pointers-in-c).

I'm pretty sure share pointers can be null, this doesn't prevent null checking. Whats more is that we will mostly use the stack, not sure what would be the advantage of shared pointer, it would actually be harmfull. Not sure what you mean with multi threaded safety of shared pointer, I think their counter is atomic but thats all. 

What would happen when the dependent goes out of scope? The destructor of the dependency is called on a stack object when it's not even out of it's scope? This would probably be undefined behavior. I agree with the singleton, we could eventually have a static class  or a namespace that manages those dependencies. I'm just not sure how you see this and how it helps here.

Reply:
My bad on this, I thought that in order to create shared pointers, you had to use std::make_shared, which takes an object and not a pointer in its constructor, making sure we do not have a null reference. You can have empty shared_pointers. :cry: . To respond to the initial comment about null checks in constructors, I would prefer using factories instead of null checks where applicable. 

For thread safety, you are right, only counters are atomic... Read some documentation too fast. We can talk later about how we manage member pointer and some other design standards. If one thing this review has shown is that we need to agree on a baseline of coding standards.